### DISCUSSION TITLE
MonoGame Game Editor like Godot (code + UI)

### DISCUSSION THREAD

**Discussion by nightblade9** [MonoGame/MonoGame#8594] [General]
TL;dr: coming from Godot, or Construct, GameMaker, etc. is there an audience for a MonoGame-based 2D game editor? Something that lets you lay out things in the UI, but also write code for everything.

Longer version: I used Godot for a few years, and used MonoGame prior to that for a few years. It would be really cool to have something like Godot's all-in-one editor in MonoGame. But it's an absolutely huge project. Is this something others want, and/or are willing to contribute to?

To build something truly encompassing of functionality (e.g. collision detection, tilemaps with auto tiles, reverb, etc.) would take a huge amount of effort, and it would have to be a community project. It would take too long for one person to build.

Would anyone be interested in building or using this? Or am I just shouting into the void?

For anyone looking for more info, the stack would be something that's built on MonoGame and friends (MonoGame Extended, Nez, etc.) depending on what users need.  The core components would be:

- Editor (the GUI editor)
- Core (the core classes/domain objects, such as sprites, labels for text, etc.)
- Game runner (takes your game project files and runs the actual game)
- Framework code (MonoGame, Nez, etc.)

You can see a sample stack in the last version of this I tried to build, [SharpForge](https://github.com/nightblade9/SharpForge/), which barely got off the ground. FlatRedBall is good, albeit it relies on external tools like Tiled etc., so the experience isn't quite the same as a built-in tool.

**Comment by SteveSuv** [comment]
This is a false proposition. I think most of MG's users are programmers who don't want to use drag-and-drop UI. Everything can be handled completely through code, rather than switching back and forth between the UI editor and code editor. If someone want to use a game engine with UI editor, maybe should choose Unity or Godot?

**Reply by tomspilman** [reply]
@SteveSuv @nightblade9 

I think there is a middle ground here.

I personally don't really believe in pure "game editors"...  you always have to figure out how to fit your game into how they want to work.

I think the MonoGame ethos is more about building custom tools for your specific game.

So what IMO would be better is a basic editor framework that provides the very basics...  cross platform windows and widgets, an undo/redo system, content management/building, build automation, etc.  None of this should be a stand alone "game IDE" or "game editor"...  it should be a designed as a "mode" you can enable in your game, so you can jump in and edit things, then jump back into gameplay.

The core would come with examples of different extensions... tile editors, scripting, level builders, etc.  The goal being provide enough basics (like XNA did for the "engine") so that a dev can take off and custom build their editor tools for their game.

It is more like how games were built before Unreal.

It could all be something the MG Foundation takes on after we get caught up on other core issues.

**Comment by ThomasFOG** [comment]
I double down on what tom highlighted. MonoGame is at core a framework to build your own engine/editor. However, MonoGame could do better in that regard, like integrating easily a GUI toolkit (like [ImGUI](https://github.com/ocornut/imgui), which is today's industry standard), modern serialization for games (like [protobuff](https://protobuf.dev/)), and have examples on how to build editors.

There's likely enough engines+editors out there. MonoGame has that special garage smell that makes it different. I believe that it would be futile to try to compete on that front.

**Comment by nightblade9** [comment]
Thank you all for the feedback, I appreciate it very much.  In brief:

> I personally don't really believe in pure "game editors"... you always have to figure out how to fit your game into how they want to work.

Same here, until I got the hang of Godot. Doing stuff in the editor and writing code makes so much more sense for certain things, like UI development.

> If someone want to use a game engine with UI editor, maybe should choose Unity or Godot?
> There's likely enough engines+editors out there.

Nah, Unity is bloated crap, and Godot has several problems. Chief among them, I can never edit the core editor or framework, because it's not in C#.

Interestingly, to contrast, I got some positive reception to this idea on social media. Albeit, positive reception is not enough for a huge undertaking like this.  I think this could be a strong competitor to something like Godot and Unity, but not something I can build solo. Maybe it'll be fun to throw together a quick prototype, though.
